browser/ipfs/ipfs_host_resolver.cc

```diff
@@ -0,0 +1,87 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include <utility>
+#include <vector>
+
+#include "brave/browser/ipfs/ipfs_host_resolver.h"
+#include "chrome/browser/net/secure_dns_config.h"
+#include "chrome/browser/net/system_network_context_manager.h"
+#include "mojo/public/cpp/bindings/receiver.h"
+#include "net/base/host_port_pair.h"
+#include "net/dns/public/dns_protocol.h"
+
+namespace ipfs {
+
+IPFSHostResolver::IPFSHostResolver(
+    network::mojom::NetworkContext* network_context,
+    const std::string& prefix)
+    : prefix_(prefix), network_context_(network_context) {
+  DCHECK(network_context);
+}
+IPFSHostResolver::~IPFSHostResolver() {}
+
+void IPFSHostResolver::Resolve(const net::HostPortPair& host,
+                               const net::NetworkIsolationKey& isolation_key,
+                               net::DnsQueryType dns_query_type,
+                               HostTextResultsCallback callback) {
+  if (!callback)
+    return;
+  if (IsRunning())
```

Comment:
It seems like we're adding complexity to this by making it reusable and I'm not sure it's 100% correct. For instance:
1. Call to resolve 1
2. Navigate to another page (callback is invalidated)
3. Call to resolve 2 (callback is now valid again)
4. Response from resolve 1 runs callback with incorrect results
Seems like it would be simpler to just subclass `ResolveHostClientImpl`, implement `OnTextResults` and reset to a new instance on navigation?

Reply:
IPFSHostResolver is only wrapper over network_context_->ResolveHost which manages resolvers.

I reset reciever on line 42, it will unbound reciever and reset actual resolver inside network service process. Third step will work with new instance each time as you proposed and it is impossible to get case from 4 step now.

I tried subclassing of ResolveHostClientImpl first but it increases necessary amount of code to achieve same result as we have here.